### PR TITLE
SLVUU-101 improve take screenshot functionality

### PR DESCRIPTION
**Pull request by vferraro-scottlogic**
There is an [open BUG](https://github.com/bubkoo/html-to-image/issues/347) tracking this issue on the html-to-image repo. The following changes are aimed at mitigating the issue with the UI freezing when taking a screenshot of a very complex layout

### Changes
- When taking the screenshot, filter DOM elements out based on role (exclude the content of every element that has role="row")
- add spinner while image is loading
- takeScreenshot function will reject with an error message if something went wrong while taking the screenshot
- SaveLayoutDialog will render: 
   - a screenshot if screenshot state is not undefined
   - an error message if screenshotErrorMessage is not undefined
   - or else a spinner 

Before

After

takeScreenshot function throws error (in this case the error is thrown because the layout is not displayed in the DOM)

### Note 
- Filtering out the row elements means we won't be displaying the content of the tables inside the screenshot which could be a positive thing if we are worried about the data being sensitive

- [x] fix unit tests

### CODE REVIEW

File: vuu-ui/packages/vuu-utils/src/screenshot-utils.ts [hunk 1]
```diff
@@ -2,22 +2,29 @@ import { toPng } from "html-to-image";
...
-  const screenshot = await toPng(node, { cacheBust: true })
-    .then((dataUrl) => {
-      return dataUrl;
+export const takeScreenshot = (node: HTMLElement): Promise<string> => {
+  return new Promise((resolve, reject) => {
+    toPng(node, {
+      cacheBust: true,
+      filter: (child) =>
+        child.nodeType === Node.TEXT_NODE ||
+        child.getAttribute("role") !== "row",
```
Comment (pling-scottlogic) [inline vuu-ui/packages/vuu-utils/src/screenshot-utils.ts:14]: These look like good filter criteria to me, but I'd like to know what @heswell thinks.

**Comment by pling-scottlogic** [comment]
Could you add some static images of equivalent screenshots before and after the change, please? It's a bit tricky to evaluate from the gifs.

**Review by pling-scottlogic — CHANGES_REQUESTED** [review]

File: vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts [hunk 2]
```diff
@@ -1,38 +1,38 @@
 import { takeScreenshot } from "../src/screenshot-utils";
 import { describe, expect, it, vi } from "vitest";
+import { expectPromiseRejectsWithError } from "./utils";
+import htmlToImage from "html-to-image";
 
-/**
```
Comment (vferraro-scottlogic) [inline vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts:4]: this comment here confused me. Not sure why we are mentioning this happy-dom package here in this test file as I don't see it being used here or in the screenshot-utils file. I have removed it for now but I am happy to add it back in if we think it's needed

File: vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts [hunk 3]
```diff
@@ -1,38 +1,38 @@
...
-
-    // We need to mock the html-to-image package because the web API operations it relies on (e.g. canvas.toDataUrl) are not available in the test environment
```
Comment (vferraro-scottlogic) [inline vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts:14]: I don't think we need to justify why we are mocking a dependency in this unit test as that is the common practice to do

File: vuu-ui/packages/vuu-shell/src/layout-management/SaveLayoutPanel.tsx [hunk 4]
```diff
@@ -125,8 +130,10 @@ export const SaveLayoutPanel = (props: SaveLayoutPanelProps) => {
               src={screenshot}
               alt="screenshot of current layout"
             />
+          ) : screenshotErrorMessage ? (
+            <Text className="screenshot">{screenshotErrorMessage}</Text>
```
[the remark this answers is not in the document]
Comment (vferraro-scottlogic) [reply vuu-ui/packages/vuu-shell/src/layout-management/SaveLayoutPanel.tsx]: turns out we dont need the style coming from that class on the error message so I have removed the class altogether. Also some of the styling from .screenshot was redundant so I have removed that as well

File: vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts (hunk 2 above)
Comment (cfisher-scottlogic) [reply vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts:4]: The `html-to-image` package uses functionality of the browser API which isn't available with the default `node` test environment, nor `jsdom`, but is available with `happy-dom`. Removing it should break the screenshot test. If it doesn't, perhaps something's changed and the default environment is fine.

File: vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts (hunk 3 above)
Comment (cfisher-scottlogic) [reply vuu-ui/packages/vuu-utils/test/screenshot-utils.test.ts:14]: You're right. I think I was getting mixed up trying to make a unit-test-integration-test crossover.

**Review by cfisher-scottlogic — APPROVED** [review]

**Review by pling-scottlogic — APPROVED** [review]
I'm happy to approve this, but I feel like the styling could be improved where we now have empty rectangles for tables. Perhaps by including a placeholder image to represent a table, and removing the scrollbars.

**Comment by vferraro-scottlogic** [comment]
[raised PR to finos](https://github.com/finos/vuu/pull/1076)

**Pull request closed without merging**